services/circleci/circleci.helpers.spec.js

```diff
@@ -0,0 +1,177 @@
+        outcome: 'success',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      {
+        outcome: 'success',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      // this failed build is part of a different workflow
+      {
+        outcome: 'failed',
+        workflows: { workflow_id: 'bbbbbbbb-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+    ]).expect('passing')
+
+    given([
+      // this workflow contains 3 builds: 2 passing builds and one which failed
+      {
+        outcome: 'success',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      {
+        outcome: 'success',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      {
+        outcome: 'failed',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      // we should summarize the status of this build as 'failed'
+    ]).expect('failed')
+
+    given([
+      // not all of the builds in the most recent wokflow are complete
+      {
+        outcome: 'success',
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      {
+        outcome: null,
+        workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      // so we should report the status of this older workflow instead
+      // because all of its builds have finished
+      {
+        outcome: 'failed',
+        workflows: { workflow_id: 'bbbbbbbb-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+      {
+        outcome: 'success',
+        workflows: { workflow_id: 'bbbbbbbb-1111-aaaa-1111-aaaaaaaaaaaa' },
+      },
+    ]).expect('failed')
+
+    expect(() =>
+      summarizeBuildsForLatestCompleteWorkflow([
+        // we have no completed workflows
+        {
+          outcome: null,
+          workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+        },
+        {
+          outcome: null,
+          workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+        },
+      ])
+    ).to.throw(Error, 'No complete workflows found')
+
+    expect(() =>
+      summarizeBuildsForLatestCompleteWorkflow([
+        {
+          outcome: 'failed',
+          workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+        },
+        // this status value is not expected
+        {
+          outcome: 'cheese',
+          workflows: { workflow_id: 'aaaaaaaa-1111-aaaa-1111-aaaaaaaaaaaa' },
+        },
+      ])
+    ).to.throw(Error, 'Found unexpected outcome')
+
+    expect(() =>
+      summarizeBuildsForLatestCompleteWorkflow([
+        // this response doesn't have workflows
+        { outcome: 'failed' },
+        { outcome: 'success' },
+      ])
+    ).to.throw(Error, "Cannot read property 'workflow_id' of undefined")
+  })
+})
+
+describe('circleci: schema validation', function() {
+  const validate = function(data, schema) {
```

Comment:
This may not be the best place to get bogged down in it, but it seems like writing tests for non-trivial validators (like this one) is something we might want to do sometimes. I definitely found I needed to write tests to get this one correct. Lets have a think about a good pattern for this.